oak_client/src/attestation.rs

```diff
@@ -0,0 +1,156 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::{anyhow, Context};
+use http::uri::Uri;
+use tokio_rustls::rustls;
+use tonic::transport::{Channel, ClientTlsConfig};
+use x509_parser::{
+    der_parser::{self, der::parse_der, oid::Oid},
+    parse_x509_certificate,
+};
+
+// Using [`Nid::NETSCAPE_COMMENT`] identifier since `rust-openssl` doesn't support custom
+// extensions yet.
+// https://github.com/sfackler/rust-openssl/issues/1411
+// https://www.alvestrand.no/objectid/2.16.840.1.113730.1.13.html
+const TEE_EXTENSION_OID: Oid<'static> = der_parser::oid!(2.16.840.1.113730.1.13);
+
+/// Creates a remotely attested TLS channel for connecting to Oak.
+pub async fn create_attested_tls_channel(
+    uri: &Uri,
+    root_tls_certificate: &[u8],
+    tee_quote: &str,
+) -> anyhow::Result<Channel> {
+    let tls_config = create_tls_attestation_config(root_tls_certificate, tee_quote)
+        .context("Couldn't create TLS config")?;
+    Channel::builder(uri.clone())
+        .tls_config(tls_config)
+        .context("Couldn't create TLS configuration")?
+        .connect()
+        .await
+        .context("Couldn't connect to Oak Application")
+}
+
+/// Create client TLS configuration with a custom X.509 certificate verifier.
+/// [`root_tls_certificate`] should contain only a single PEM encoded certificate.
+/// https://tools.ietf.org/html/rfc1421
+pub fn create_tls_attestation_config(
+    root_tls_certificate: &[u8],
+    tee_quote: &str,
+) -> anyhow::Result<ClientTlsConfig> {
+    let mut config = rustls::ClientConfig::new();
+
+    // Enable HTTP/2.
+    const ALPN_H2: &str = "h2";
+    config.set_protocols(&[Vec::from(&ALPN_H2[..])]);
+
+    // Add root TLS certificate.
+    let mut cc_reader = std::io::BufReader::new(&root_tls_certificate[..]);
+    let certs = rustls::internal::pemfile::certs(&mut cc_reader)
+        .map_err(|error| anyhow!("Couldn't parse TLS certificate: {:?}", error))?;
+    if certs.len() != 1 {
+        return Err(anyhow!("Only a single PEM certificate should be provided"));
+    }
+    config
+        .root_store
+        .add(&certs[0])
+        .context("Couldn't add root certificate")?;
+
+    // Set custom X.509 certificate verifier.
+    config
+        .dangerous()
+        .set_certificate_verifier(std::sync::Arc::new(RemoteAttestationVerifier::new(
+            tee_quote,
+        )));
+
+    Ok(ClientTlsConfig::new().rustls_client_config(config))
+}
+
+/// Custom X.509 certificate verifier that checks TEE quotes in the X.509 TEE extension.
+struct RemoteAttestationVerifier {
+    tee_quote: String,
+}
+
+impl RemoteAttestationVerifier {
+    fn new(tee_quote: &str) -> Self {
+        Self {
+            tee_quote: tee_quote.to_string(),
+        }
+    }
+
+    fn check_tee_quote(&self, extension: &str) -> anyhow::Result<()> {
+        let pattern = r"^Quote:(.*)$";
```

Comment:
Not sure I get what's going on here, why not use JSON or something that can be parsed properly? It does not seem it would make the code much longer, and I think it is probably more similar to what we would expect in the future? Or just make the entire field the quote, without the `Quote` prefix, for now. Or is the prefix needed for something?